File: python-assignments/Linked list/linked-list.py
# Node structure
class Node:
    def __init__(self, value):
        self.data = value
        self.next = None

# Linked list
class LinkedList:
    def __init__(self):
        self.head = None
        self.size = 0

# insert at index
def insert_at(lst, index, value):
    if index < 0 or index > lst.size:
        print("Error: Invalid index.")
        return
    new_node = Node(value)
    if index == 0:  # insert at start
        new_node.next = lst.head
        lst.head = new_node
    else:  # insert middle/end
        current = lst.head
        for _ in range(index - 1):
            current = current.next
        new_node.next = current.next
        current.next = new_node
    lst.size += 1

# delete at index
def delete_at(lst, index):
    if lst.size == 0:
        print("Error: List is empty.")
        return
    if index < 0 or index >= lst.size:
        print("Error: Invalid index.")
        return
    if index == 0:  # delete first
        temp = lst.head
        lst.head = lst.head.next
    else:  # delete middle/end
        current = lst.head
        for _ in range(index - 1):
            current = current.next
        temp = current.next
        current.next = temp.next
    temp.next = None  # optional
    lst.size -= 1

# print list
def traverse(lst):
    current = lst.head
    if current is None:
        print("[empty]")
        return
    while current is not None:
        print(current.data, end=" ")
        current = current.next
    print()

# main
if __name__ == "__main__":
    lst = LinkedList()
    insert_at(lst, 0, 10)
    insert_at(lst, 1, 20)
    insert_at(lst, 1, 15)
    insert_at(lst, 3, 25)

    print("List after insertions: ", end="")
    traverse(lst)

    delete_at(lst, 1)
    print("List after deleting index 1: ", end="")
    traverse(lst)

    delete_at(lst, 0)
    print("List after deleting index 0: ", end="")
    traverse(lst)

File: pascal-assignments/Linked-list/linked-list.pas
program LinkedListDemo;

type
  NodePtr = ^Node;

  Node = record
    data : integer;
    next : NodePtr;
  end;

var
  head : NodePtr;
  size : integer;

{ insert at index }
procedure InsertAt(index, value : integer);
var
  newNode, current : NodePtr;
  i : integer;
begin
  if (index < 0) or (index > size) then
  begin
    writeln('Error: Invalid index');
    exit;
  end;

  new(newNode);
  newNode^.data := value;
  newNode^.next := nil;

  if index = 0 then
  begin
    newNode^.next := head;
    head := newNode;
  end
  else
  begin
    current := head;
    for i := 0 to index - 2 do
      current := current^.next;

    newNode^.next := current^.next;
    current^.next := newNode;
  end;

  size := size + 1;
end;

{ delete at index }
procedure DeleteAt(index : integer);
var
  current, temp : NodePtr;
  i : integer;
begin
  if size = 0 then
  begin
    writeln('Error: List is empty');
    exit;
  end;

  if (index < 0) or (index >= size) then
  begin
    writeln('Error: Invalid index');
    exit;
  end;

  if index = 0 then
  begin
    temp := head;
    head := head^.next;
  end
  else
  begin
    current := head;
    for i := 0 to index - 2 do
      current := current^.next;

    temp := current^.next;
    current^.next := temp^.next;
  end;

  dispose(temp);
  size := size - 1;
end;

{ print list }
procedure Traverse;
var
  current : NodePtr;
begin
  if head = nil then
  begin
    writeln('[empty]');
    exit;
  end;

  current := head;
  while current <> nil do
  begin
    write(current^.data, ' ');
    current := current^.next;
  end;
  writeln;
end;

begin
  { start }
  head := nil;
  size := 0;

  InsertAt(0, 10);
  InsertAt(1, 20);
  InsertAt(1, 15);
  InsertAt(3, 25);

  write('List after insertions: ');
  Traverse;

  DeleteAt(1);
  write('List after deleting index 1: ');
  Traverse;

  DeleteAt(0);
  write('List after deleting index 0: ');
  Traverse;
end.
